File: app/src/main/java/com/nikola/jakshic/dagger/viewModel/SearchViewModel.java
package com.nikola.jakshic.dagger.viewModel;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.nikola.jakshic.dagger.AppExecutors;
import com.nikola.jakshic.dagger.data.local.SearchHistoryDao;
import com.nikola.jakshic.dagger.model.Player;
import com.nikola.jakshic.dagger.model.SearchHistory;
import com.nikola.jakshic.dagger.repository.PlayerRepository;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class SearchViewModel extends ViewModel {

    private MutableLiveData<List<Player>> playerList;
    private MutableLiveData<List<SearchHistory>> historyList;
    private PlayerRepository repository;
    private SearchHistoryDao searchHistoryDao;
    private MutableLiveData<Boolean> loading;
    private AppExecutors executor;

    @Inject
    public SearchViewModel(AppExecutors executor, PlayerRepository repository, SearchHistoryDao searchHistoryDao) {
        this.repository = repository;
        this.executor = executor;
        this.searchHistoryDao = searchHistoryDao;
        playerList = new MutableLiveData<>();
        historyList = new MutableLiveData<>();
        loading = new MutableLiveData<>();
        loading.setValue(false);
    }

    public void getAllQueries() {
        executor.diskIO().execute(() -> {
            List<SearchHistory> allQueries = searchHistoryDao.getAllQueries();
            historyList.postValue(allQueries);
        });

    }

    public MutableLiveData<List<SearchHistory>> getSearchHistory() {
        return historyList;
    }

    public void getQueries(String query) {
        executor.diskIO().execute(() -> {
            List<SearchHistory> list = searchHistoryDao.getQuery(query);
            historyList.postValue(list);
        });

    }

    public void saveQuery(SearchHistory searchHistory) {
        executor.diskIO().execute(() -> {
            searchHistoryDao.insertQuery(searchHistory);
        });
    }

    public MutableLiveData<List<Player>> getPlayers() {
        return playerList;
    }

    public void fetchPlayers(String name) {
        repository.fetchPlayers(playerList, loading, name);
    }

    public MutableLiveData<Boolean> isLoading() {
        return loading;
    }

    public void clearList() {
        playerList.setValue(new ArrayList<>());
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/viewModel/SearchViewModel.kt
package com.nikola.jakshic.dagger.viewModel

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.nikola.jakshic.dagger.Status
import com.nikola.jakshic.dagger.data.local.SearchHistoryDao
import com.nikola.jakshic.dagger.model.Player
import com.nikola.jakshic.dagger.model.SearchHistory
import com.nikola.jakshic.dagger.repository.PlayerRepository
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

class SearchViewModel @Inject constructor(
        private val dao: SearchHistoryDao,
        private val repository: PlayerRepository
) : ViewModel() {

    val playerList = MutableLiveData<List<Player>>()
    val historyList = MutableLiveData<List<SearchHistory>>()
    val status = MutableLiveData<Status>()
    private val disposables = CompositeDisposable()

    fun getAllQueries() {
        disposables.add(Single
                .fromCallable(dao::getAllQueries)
                .subscribeOn(Schedulers.io())
                .subscribe(historyList::postValue))
    }

    fun getQueries(query: String) {
        disposables.add(Single
                .fromCallable { dao.getQuery(query) }
                .subscribeOn(Schedulers.io())
                .subscribe(historyList::postValue))
    }

    fun saveQuery(item: SearchHistory) {
        disposables.add(Completable
                .fromAction { dao.insertQuery(item) }
                .subscribeOn(Schedulers.io())
                .subscribe())
    }

    fun fetchPlayers(name: String) {
        disposables.add(repository.fetchPlayers(playerList, status, name))
    }

    fun clearList() {
        playerList.value = ArrayList()
    }

    override fun onCleared() {
        disposables.clear()
    }
}